[FSDP] Option for eval in fp32/bf16

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104682

In https://github.com/pytorch/pytorch/pull/97645 and some follow up diffs, we made FSDP run in full precision in eval mode, even if mixed precision was specified.

However, this is probably not the best idea and we should provide a flag for users to have control over this a bit more. Adding an env var FSDP_FULL_PREC_IN_EVAL and defaulting it to off, users who want to run eval in fp32 can toggle this before wrapping model in FSDP:

os.environ["FSDP_FULL_PREC_IN_EVAL"] = "1"

Verified that unittests, APS workflow, TNT workloads can run eval appropriately with this change.

Differential Revision: [D47246556](https://our.internmc.facebook.com/intern/diff/D47246556/)

rohan-varma committed:
[FSDP] Option for eval in fp32/bf16

In https://github.com/pytorch/pytorch/pull/97645 and some follow up diffs, we made FSDP run in full precision in eval mode, even if mixed precision was specified.

However, this is probably not the best idea and we should provide a flag for users to have control over this a bit more. Adding an env var FSDP_FULL_PREC_IN_EVAL and defaulting it to off, users who want to run eval in fp32 can toggle this before wrapping model in FSDP:

os.environ["FSDP_FULL_PREC_IN_EVAL"] = "1"

Verified that unittests, APS workflow, TNT workloads can run eval appropriately with this change.

Differential Revision: [D47246556](https://our.internmc.facebook.com/intern/diff/D47246556/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -251,7 +252,10 @@ def _share_state_and_init_handle_attrs(
     root_state._all_handles = root_state._exec_order_data.all_handles  # share reference
     root_state._device_mesh = _init_device_mesh(root_state)
-
+    import os
+    root_state._full_prec_in_eval = (
+        os.environ.get(_FSDP_FULL_PREC_IN_EVAL, "") == "1"
+    )
```
We should move this to construction time. For example, we can check the new env var after this:
https://github.com/pytorch/pytorch/blob/adf1405909a3136b51be17992a8def4297764022/torch/distributed/fsdp/_init_utils.py#L393-L397

awgu commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -87,6 +87,9 @@
 
+# Env var toggling whether when model is in .eval() mode, should we run in fp32
+# or the reduced precision.
+_FSDP_FULL_PREC_IN_EVAL = "FSDP_FULL_PREC_IN_EVAL"
```
nit: Should the naming be `_FSDP_EVAL_IN_FULL_PREC = "FSDP_FULL_PREC_IN_EVAL"` or `_FSDP_USE_FULL_PREC_IN_EVAL = "FSDP_USE_FULL_PREC_IN_EVAL"`?

My reasoning here is that you save this to an attribute `_full_prec_in_eval: bool`, so I would read it like "if full precision in eval". I read it more naturally as "if eval in full precision" or "if use full precision in eval".

rohan-varma replied:
sounds good!

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -872,21 +900,27 @@ def test_full_precision_in_eval_buffers(self):
                 CUDAInitMode.CUDA_BEFORE,
                 {"mixed_precision": mp_config},
             )
+            for fsdp_module in FSDP.fsdp_modules(model):
+                mp = fsdp_module.mixed_precision
```
Where is `mp` used? Why do we take the last one of the FSDP modules?

awgu replied:
Why is this resolved? :)

rohan-varma replied:
oops, I deleted this locally but somehow didn't push that change.

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -339,14 +339,17 @@ def init(
-                TransformerWithSharedParams(
+            m = TransformerWithSharedParams(
                     tformer_pg, cuda_init_mode, add_bn, deterministic
-                ),
+                )
+            fsdp_model = FSDP(
+                m,
                 fsdp_pg,
                 auto_wrap_policy=auto_wrap_policy,
                 **fsdp_kwargs,
             )
+            for fsdp_module in FSDP.fsdp_modules(fsdp_model):
+                mp = fsdp_module.mixed_precision
```
Could you explain this change?

rohan-varma replied:
Debugging vestige, not needed

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -761,41 +766,54 @@ def test_eval_root_cast_inputs(self):
+        this.
         """
 
+        reduce_dtype = torch.float16
```
nit: Curious why you named this `reduce_dtype` when we use it for `param_dtype`, `reduce_dtype`, and `buffer_dtype`? Maybe name it `low_prec_dtype`?

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -761,41 +766,54 @@ def test_eval_root_cast_inputs(self):
-            def forward(self, x):
-                assert x.dtype == torch.float32, f"Expected fp32, got {x.dtype}"
+            def forward(self, x, expect_full_prec_in_eval):
+                if expect_full_prec_in_eval:
+                    assert x.dtype == torch.float32, f"Expected fp32, got {x.dtype}"
+                else:
+                    assert x.dtype == reduce_dtype, f"Expected {reduce_dtype}, got {x.dtype}"
```
I guess the pain here is that t he `self` arg to `forward()` shadows the test class `self`, so it is hard to use `self.assertEquals()`.

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -860,9 +883,14 @@ def test_full_precision_in_eval_buffers(self):
+            # buffer computation takes place in the right precision.
             for buf in fsdp_model.buffers():
-                self.assertEqual(_BUFFER_ORIG_DTYPE, buf.dtype)
+                if full_prec_in_eval:
```
nit: You do `expected_dtype = ... if full_prec_in_eval else ...` above. You could do it here too for slight improvement in readability.

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -860,9 +883,14 @@ def test_full_precision_in_eval_buffers(self):
             fsdp_model.eval()
             fsdp_model((inp, self, fsdp_model, mp_config, torch.float32))
-
+            # TODO: this test would be more robust if the buffer dtype was
```
Make this a bootcamp/onboarding task? :)

rohan-varma replied:
https://github.com/pytorch/pytorch/issues/104740

awgu reviewed (commented):
This change makes sense to me. I mainly want to double check (1) env var / attribute name  and (2) moving the initialization of the attribute to construction time instead of lazy initialization time.

rohan-varma committed:
Update on "[FSDP] Option for eval in fp32/bf16"


In https://github.com/pytorch/pytorch/pull/97645 and some follow up diffs, we made FSDP run in full precision in eval mode, even if mixed precision was specified.

However, this is probably not the best idea and we should provide a flag for users to have control over this a bit more. Adding an env var FSDP_FULL_PREC_IN_EVAL and defaulting it to off, users who want to run eval in fp32 can toggle this before wrapping model in FSDP:

os.environ["FSDP_FULL_PREC_IN_EVAL"] = "1"

Verified that unittests, APS workflow, TNT workloads can run eval appropriately with this change.

Differential Revision: [D47246556](https://our.internmc.facebook.com/intern/diff/D47246556/)

[ghstack-poisoned]

awgu reviewed (approved):
LGTM!

A minor nit is if you want, you could change the `full_prec_in_eval` in test_fsdp_mixed_precsion.py to `use_full_prec_in_eval` to match the new env var name (as a simple find & replace).

rohan-varma committed:
Update on "[FSDP] Option for eval in fp32/bf16"


In https://github.com/pytorch/pytorch/pull/97645 and some follow up diffs, we made FSDP run in full precision in eval mode, even if mixed precision was specified.

However, this is probably not the best idea and we should provide a flag for users to have control over this a bit more. Adding an env var FSDP_FULL_PREC_IN_EVAL and defaulting it to off, users who want to run eval in fp32 can toggle this before wrapping model in FSDP:

os.environ["FSDP_FULL_PREC_IN_EVAL"] = "1"

Verified that unittests, APS workflow, TNT workloads can run eval appropriately with this change.

Differential Revision: [D47246556](https://our.internmc.facebook.com/intern/diff/D47246556/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Option for eval in fp32/bf16"


In https://github.com/pytorch/pytorch/pull/97645 and some follow up diffs, we made FSDP run in full precision in eval mode, even if mixed precision was specified.

However, this is probably not the best idea and we should provide a flag for users to have control over this a bit more. Adding an env var FSDP_FULL_PREC_IN_EVAL and defaulting it to off, users who want to run eval in fp32 can toggle this before wrapping model in FSDP:

os.environ["FSDP_FULL_PREC_IN_EVAL"] = "1"

Verified that unittests, APS workflow, TNT workloads can run eval appropriately with this change.

Differential Revision: [D47246556](https://our.internmc.facebook.com/intern/diff/D47246556/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -606,6 +606,7 @@ def test_homogeneous_attributes(self):
         all_attr_name_and_values = [
             ("_use_orig_params", False, True),
             ("limit_all_gathers", False, True),
+            ("_use_full_prec_in_eval", False, True),
```
@awgu this is an attr that should be set uniformly for all handles, but isn't a direct ctor input. So I added the below logic to test, LMK if shoudl do something different.

rohan-varma commented:
@pytorchbot merge -f "CI done"
